docs/source/en/guides/upload.md

```diff
@@ -431,6 +431,50 @@ In addition to [`upload_file`] and [`upload_folder`], the following functions al
 
 For more detailed information, take a look at the [`HfApi`] reference.
 
+### Preupload LFS files before commit
+
+In some cases, you might want to upload huge files to S3 **before** making the commit call. For example, if you are
+committing a dataset in several shards that are generated in-memory, you would need to upload the shards one by one
+to avoid a out-of-memory issue. A solution is to upload each shard as a separate commit on the repo. While being
+perfectly valid, this solution has the drawback of potentially messing the git history by generating a tens of commits.
+To overcome this issue, you can upload your files one by one to S3 and then create a single commit at the end. This
+is possible using [`preupload_lfs_files`] in combination with [`create_commit`].
+
+<Tip warning={true}>
+
+This is a power-user method. Directly using [`upload_file`], [`upload_folder`] or [`create_commit`] instead of handling
+the low-level logic of pre-uploading files is the way to go in the vast majority of cases. The main caveat of
+[`preupload_lfs_files`] is that until the commit is actually made, the upload files are not accessible on the repo on
+the Hub. If you have a question, feel free to ping us on our Discord or in a Github issue.
+
+</Tip>
+
+Here is a simple example illustrating how to pre-upload files:
+
+```py
+>>> from huggingface_hub import CommitOperationAdd, preupload_lfs_files, create_commit, create_repo
+
+>>> repo_id = create_repo("test_preupload").repo_id
+
+>>> operations = [] # List of all `CommitOperationAdd` objects that will be generated
+>>> for i in range(5):
+...     content = ... # generate binary content
+...     addition = CommitOperationAdd(path_in_repo=f"shard_{i}_of_5.bin", path_or_fileobj=content)
+...     preupload_lfs_files(repo_id, additions=[addition])
+...     operations.append(addition)
+
+# Create commit
+>>> create_commit(repo_id, operations=operations, commit_message="Commit all shards")
+```
```

Comment:
What happens if there is a failure on one of the operations before the commit is created?

Reply:
If there is a failure in one of the uploads, the script will crash (if not in a try/except). If the script is restarted, already uploaded files will not need to be re-uploaded (`preupload_lfs_files` will be _instant_). However, as long as the `create_commit` operation is not completed, the files are not accessible to anyone and cannot be seen on the repo on the Hub. 

Also there was a question at some point to garbage-collect the untracked files in S3 after some time (after 24h?) cc @Pierrci. I think it's not enabled yet but that would mean that if the user waits too long before creating the commit then it's lost.

Reply:
Is there a way to check if a certain file has been preuploaded based on its name ?

That would help implementing `push_to_hub` resuming if it crashed mid-way

Reply:
> Is there a way to check if a certain file has been preuploaded based on its name ? Or it requires the hash ?

It requires the hash unfortunately. Filenames are just convenient aliases saved in git but what matters on S3 are the uniqueness of the files (i.e. based on hash). For context, until the `create_commit` step, the filename is not even provided to the Hub.

Reply:
Ok I see. I guess we can do a commit every N files to allow resuming from there

Reply:
@lhoestq Yes indeed.
But in general, what takes the most time in `push_to_hub`? Is it generating + hashing the shards or uploading them? Because if generate+hash takes 5% of the upload time for instance, resuming from a failed `push_to_hub` should still be very fast.

Reply:
Our current "resuming" logic is to generate the shards and compute their "fingerprint" to check if they are already present in the repo (basically, identical to hashing), so "resuming from a failed `push_to_hub`" should be fine, as this step is pretty fast (I don't remember anyone complaining about it being slow)